File: objectiveai-sdk-js/src/daemon/fileTree.ts
/**
 * Materialized consumer of the cli daemon's
 * `/laboratories/{id}/filetree` endpoint — the JS mirror of the Rust
 * SDK's `daemon::file_tree::FileTree`, folded with the same event
 * semantics (`FileTreeEvent::apply` in `wire.rs`).
 *
 * The fold here is deliberately PATH-COPYING IMMUTABLE (the repo's
 * merge-system identity convention, unlike the Rust in-place fold):
 * every applied event produces a NEW root array, shallow-copying only
 * the spine from the root to the changed node — every untouched
 * sibling subtree keeps its object identity. React consumers can
 * therefore `memo` per-node components and re-render exactly the
 * changed region.
 *
 * One listener = one connection: when the stream closes the view
 * freezes at its last state; reconnect is the caller's loop — mint a
 * new listener from the client. Unparseable events are skipped
 * (forward compat).
 */

import type {
  LaboratoriesFiletreeFileTreeEvent,
  LaboratoriesFiletreeFileTreeNode,
} from "../laboratories/filetree/index";
import { openStream, type ClientMode } from "./mode";

type FileTreeNode = LaboratoriesFiletreeFileTreeNode;
type FileTreeEvent = LaboratoriesFiletreeFileTreeEvent;

/**
 * THE fold — the TS port of the Rust SDK's `FileTreeEvent::apply`,
 * path-copying instead of mutating: `snapshot` replaces the whole
 * child set; `upserted` rebuilds the spine down `path` and
 * replaces-or-appends the leaf by name (missing middle segments are
 * synthesized as empty directories, non-directories in the way are
 * replaced); `removed` rebuilds the spine and drops the leaf (a
 * missing path is a no-op returning the SAME root — no spurious
 * identity change). Idempotent, so at-least-once delivery is safe.
 */
export function applyFileTreeEvent(
  root: FileTreeNode[],
  event: FileTreeEvent,
): FileTreeNode[] {
  switch (event.type) {
    case "snapshot":
      return event.children;
    case "upserted":
      if (event.path.length === 0) return root;
      return upsertAt(root, event.path, 0, event.node);
    case "removed":
      if (event.path.length === 0) return root;
      return removeAt(root, event.path, 0);
    default:
      // Unknown event type — forward compat, no change.
      return root;
  }
}

/** Copy `children` with `node` installed at `path[depth..]`,
 * spine-copying directories on the way down. */
function upsertAt(
  children: FileTreeNode[],
  path: string[],
  depth: number,
  node: FileTreeNode,
): FileTreeNode[] {
  const name = path[depth];
  const next = [...children];
  const i = next.findIndex((child) => child.name === name);
  if (depth === path.length - 1) {
    if (i >= 0) next[i] = node;
    else next.push(node);
    return next;
  }
  const existing = i >= 0 ? next[i] : undefined;
  const directory: FileTreeNode =
    existing !== undefined && existing.type === "directory"
      ? {
          ...existing,
          children: upsertAt(existing.children, path, depth + 1, node),
        }
      : {
          // Missing middle (or a non-directory in the way) — a bare
          // synthesized directory, exactly like the Rust fold.
          type: "directory",
          name,
          children: upsertAt([], path, depth + 1, node),
        };
  if (i >= 0) next[i] = directory;
  else next.push(directory);
  return next;
}

/** Copy `children` with the node at `path[depth..]` removed,
 * spine-copying on the way down. Returns the SAME array when the
 * path doesn't resolve — no change, no identity churn. */
function removeAt(
  children: FileTreeNode[],
  path: string[],
  depth: number,
): FileTreeNode[] {
  const name = path[depth];
  const i = children.findIndex((child) => child.name === name);
  if (i < 0) return children;
  if (depth === path.length - 1) {
    const next = [...children];
    next.splice(i, 1);
    return next;
  }
  const existing = children[i];
  if (existing.type !== "directory") return children;
  const nextChildren = removeAt(existing.children, path, depth + 1);
  if (nextChildren === existing.children) return children;
  const next = [...children];
  next[i] = { ...existing, children: nextChildren };
  return next;
}

/**
 * The materialized `/laboratories/{id}/filetree` view — minted by
 * {@link Client.fileTree}; resolves only once the stream has OPENED.
 *
 * ```ts
 * const tree = await client.fileTree("my-lab");
 * tree.children(); // FileTreeNode[] | null (null until the first frame)
 * await tree.subscribe(); // resolves on the next change
 * ```
 */
export class FileTree {
  #abort: AbortController;
  #closed = false;
  /** The folded root child set — `null` until the first frame lands
   * (an unknown/not-yet-started laboratory sends an EMPTY snapshot,
   * which is a real state, distinct from "still connecting"). */
  #root: FileTreeNode[] | null = null;
  /** Resolvers for the in-flight {@link subscribe} promises. */
  #waiters = new Set<() => void>();

  private constructor(abort: AbortController) {
    this.#abort = abort;
  }

  /** @internal — minted by `Client.fileTree`. `laboratoryId` is the
   * raw laboratory id; the optional `(machine, machineState)` pair
   * pins the exact owning host (both or neither). */
  static async _connect(
    mode: ClientMode,
    laboratoryId: string,
    options?: { machine?: string; machineState?: string },
  ): Promise<FileTree> {
    const abort = new AbortController();
    const events = await openStream(mode, abort.signal, {
      route: `/laboratories/${laboratoryId}/filetree`,
      query: {
        machine: options?.machine,
        machine_state: options?.machineState,
      },
      viewerCommand: "daemon_laboratory_filetree",
      viewerArgs: {
        id: laboratoryId,
        machine: options?.machine,
        machineState: options?.machineState,
      },
    });
    const tree = new FileTree(abort);
    void tree.#pump(events);
    return tree;
  }

  /** Read the SSE stream, feeding each event to {@link #onFrame} until
   * it ends (or is aborted); then settle as closed. */
  async #pump(events: AsyncGenerator<string>): Promise<void> {
    try {
      for await (const data of events) {
        let frame: unknown;
        try {
          frame = JSON.parse(data);
        } catch {
          continue;
        }
        this.#onFrame(frame);
      }
    } catch {
      // Aborted or transport error — fall through to close.
    } finally {
      this.#onClose();
    }
  }

  /** Whether the connection has closed (the view is frozen). */
  get closed(): boolean {
    return this.#closed;
  }

  /** Drop the connection: the view freezes and any pending
   * {@link subscribe} resolves. */
  close(): void {
    if (this.#closed) return;
    this.#abort.abort();
    this.#onClose();
  }

  /** The folded root child set — `null` until the first frame. The
   * array (and every unchanged subtree in it) is identity-stable
   * across events; only changed spines get new objects. */
  children(): FileTreeNode[] | null {
    return this.#root;
  }

  /** Resolves on the next change applied to the state. A fresh call
   * waits for the FIRST change after it is made — loop with the
   * {@link children} read. Resolves immediately if already closed. */
  subscribe(): Promise<void> {
    if (this.#closed) return Promise.resolve();
    return new Promise<void>((resolve) => {
      this.#waiters.add(resolve);
    });
  }

  #onFrame(frame: unknown): void {
    if (typeof frame !== "object" || frame === null) return;
    const kind = (frame as { type?: unknown }).type;
    if (kind !== "snapshot" && kind !== "upserted" && kind !== "removed") {
      return;
    }
    this.#root = applyFileTreeEvent(
      this.#root ?? [],
      frame as FileTreeEvent,
    );
    const waiters = [...this.#waiters];
    this.#waiters.clear();
    for (const wake of waiters) wake();
  }

  #onClose(): void {
    if (this.#closed) return;
    this.#closed = true;
    const waiters = [...this.#waiters];
    this.#waiters.clear();
    for (const wake of waiters) wake();
  }
}

File: objectiveai-sdk-rs/src/daemon/file_tree.rs
//! Materialized consumer of a live file-tree SSE stream — a
//! laboratory MCP's own `/filetree` endpoint
//! ([`FileTree::laboratory`]) or a CLI daemon's
//! `/laboratories/{id}/filetree` proxy
//! ([`Client::file_tree`](crate::daemon::Client::file_tree)); both
//! speak the identical wire contract.
//!
//! [`FileTree`] is NOT a raw event stream — it connects once, then
//! folds every incoming [`FileTreeEvent`] into an in-memory,
//! self-updating recursive tree (the watched root's
//! [`FileTreeNode`] children): a
//! [`Snapshot`](FileTreeEvent::Snapshot) replaces the whole child set,
//! [`Upserted`](FileTreeEvent::Upserted) inserts/replaces one node at
//! its path (a directory node carries its whole subtree), and
//! [`Removed`](FileTreeEvent::Removed) drops the node at its path.
//!
//! Ways to observe it:
//! - [`tree`](FileTree::tree) — async snapshot of the current tree.
//! - [`subscribe`](FileTree::subscribe) — async, blocks until the next
//!   change.
//! - [`changes`](FileTree::changes) — the raw change-counter receiver,
//!   for race-free condition waits.
//!
//! The machine/machine_state query the daemon route accepts is not
//! surfaced here (no consumer); add when needed.
//!
//! One [`FileTree`] = one connection: the internal pump runs until the
//! laboratory closes the stream; after that the view is frozen at its
//! last state. Dropping it aborts the pump. Reconnection is the
//! caller's loop — mint a new [`FileTree`].

use std::sync::Arc;

use futures::StreamExt;
use reqwest_eventsource::{Event, RequestBuilderExt};
use tokio::sync::{Mutex, watch};

use crate::daemon::Error;
use crate::laboratories::filetree::{FileTreeEvent, FileTreeNode};

/// The shared inner state, held by both the [`FileTree`] handle and
/// its pump task.
struct Shared {
    /// The tree itself: the watched root's child nodes, kept live by
    /// folding each event.
    state: Mutex<Vec<FileTreeNode>>,
    /// A monotonically-bumped change counter. Each applied event bumps
    /// it, waking every [`subscribe`](FileTree::subscribe) waiter.
    changes: watch::Sender<u64>,
}

/// The materialized `/filetree` view — see the module docs. Minted by
/// [`Client::file_tree`](crate::daemon::Client::file_tree) (daemon
/// flavor) or [`FileTree::laboratory`]; returned only once the stream
/// has OPENED. Dropping it aborts the background pump.
pub struct FileTree {
    shared: Arc<Shared>,
    pump: tokio::task::JoinHandle<()>,
}

impl FileTree {
    /// The daemon flavor — minted by
    /// [`Client::file_tree`](crate::daemon::Client::file_tree). Opens
    /// the daemon's `/laboratories/{id}/filetree` proxy (the RAW
    /// laboratory id — ids forbid `/`, so the URL is safe by
    /// construction), awaiting the open frame, and starts the pump.
    /// Same wire contract as the laboratory flavor — the daemon
    /// re-emits the identical events from its own materialized state,
    /// always serving the laboratory's workspace (its cwd).
    pub(crate) async fn connect_daemon(
        client: &crate::daemon::Client,
        laboratory_id: &str,
    ) -> Result<FileTree, Error> {
        let source = client
            .open_sse(&format!("/laboratories/{laboratory_id}/filetree"))
            .await?;
        Ok(Self::start(source))
    }

    /// The LABORATORY flavor: dial a laboratory's OWN `/filetree` (no
    /// auth — loopback-only lab server), optionally scoped to `path`
    /// (an absolute path inside the container), from its base HTTP
    /// address (e.g. `http://127.0.0.1:<port>` — the loopback-published
    /// container port; the port itself is the trust boundary). Awaits
    /// the open frame, then starts the pump.
    pub async fn laboratory(
        base_url: &str,
        path: Option<&str>,
    ) -> Result<FileTree, Error> {
        let mut request = reqwest::Client::new()
            .get(format!("{}/filetree", base_url.trim_end_matches('/')))
            .header("Accept", "text/event-stream");
        if let Some(path) = path {
            request = request.query(&[("path", path)]);
        }
        let mut source = request.eventsource()?;
        // reqwest-eventsource yields `Open` first on a successful
        // response; an error (invalid status included) or immediate
        // end means the stream never opened.
        match source.next().await {
            Some(Ok(Event::Open)) => {}
            Some(Err(e)) => return Err(Error::Open(e)),
            Some(Ok(Event::Message(_))) | None => return Err(Error::Closed),
        }
        Ok(Self::start(source))
    }

    /// Build the shared state and spawn the pump over an
    /// already-opened stream.
    fn start(source: reqwest_eventsource::EventSource) -> FileTree {
        let shared = Arc::new(Shared {
            state: Mutex::new(Vec::new()),
            changes: watch::channel(0u64).0,
        });
        let pump = tokio::spawn(pump(source, shared.clone()));
        FileTree { shared, pump }
    }

    /// Snapshot the current tree — the watched root's child nodes.
    /// Empty before the first snapshot has been folded.
    pub async fn tree(&self) -> Vec<FileTreeNode> {
        self.shared.state.lock().await.clone()
    }

    /// Block until the next change is applied. A fresh call waits for
    /// the FIRST change after it is made, so a change that lands
    /// between a preceding [`tree`](Self::tree) read and this
    /// call is not observed by it — pair with the read in a loop, or
    /// hold a [`changes`](Self::changes) receiver for race-free waits.
    pub async fn subscribe(&self) {
        let mut rx = self.shared.changes.subscribe();
        let _ = rx.changed().await;
    }

    /// The raw change-counter receiver — for RACE-FREE condition
    /// waits: hold ONE receiver across iterations of a
    /// check-then-await loop, and an event landing between the check
    /// and the await still resolves the next `changed()`.
    pub fn changes(&self) -> watch::Receiver<u64> {
        self.shared.changes.subscribe()
    }
}

impl Drop for FileTree {
    fn drop(&mut self) {
        self.pump.abort();
    }
}

/// Read SSE messages, fold each [`FileTreeEvent`] into `shared.state`,
/// and bump the change counter. Runs until the stream closes. Parse
/// errors and non-message events are skipped; a transport error ends
/// the pump (the caller rebuilds — reconnection is not automatic).
async fn pump(mut source: reqwest_eventsource::EventSource, shared: Arc<Shared>) {
    while let Some(event) = source.next().await {
        match event {
            Ok(Event::Open) => continue,
            Ok(Event::Message(message)) => {
                let Ok(event) = serde_json::from_str::<FileTreeEvent>(&message.data)
                else {
                    // Skip a frame we can't parse rather than tearing down.
                    continue;
                };
                {
                    let mut state = shared.state.lock().await;
                    event.apply(&mut state);
                }
                shared.changes.send_modify(|version| {
                    *version = version.wrapping_add(1);
                });
            }
            // Transport error (including the stream closing): stop.
            Err(_) => break,
        }
    }
}
